### PR TITLE
Restructure binning subworkflow in preparation for aDNA workflow and extended binning

### PR DESCRIPTION
**Pull request by jfy133**
This PR restructures the binning subworkflows to make it more generic. This is in preparation for adds maxbin2 binning as an additional binner, as well as for the aDNA workflow.

For the end user, in principle, nothing should change (other than slight re-structuring of the `GenomeBinning/Metabat2` folder), but the results should be the same. **However** please check this on some more datasets - I only tested (note: extensively) the test dataset, but someone should check with their own dataset they are familiar with.

- This splits up the binning modules and related into more discrete steps in the workflow. 
- It also replaces local modules with the latest nf-core DSL2 v2 syntax. 
- It tweaks some of the custom python scripts to accept gzipped input as per nf-core modules standard

> :warning: this does mean we have three types of module syntax: custom local, nf-core v1, nf-core v2, but I didn't want to mess with too much stuff in one go when I'm not super familiar. `modules.conf` is particular messy in this regard.

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/mag/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/mag _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by d4straub** [comment]
Hi, would it be helpful if I integrate the template update 2.2 next week or so or would that rather mess with your efforts in that PR (merge conflicts)? I am asking because I might be able to invest some time into this (and the template update needs to be done eventually anyway), but I can use the time also working on other tasks and postpone the template update.

**Comment by jfy133** [comment]
@d4straub I'll reply on slack, slightly easier ;)

**Review by d4straub — COMMENTED** [review]
Great work! I went through the code and have a couple of comments.
Just to clarify, MaxBin2 isnt used yet at all, but the module is already added?
I have not run/tested the pipeline yet, going to do that tomorrow.

File: conf/modules.config [hunk 1]
```diff
@@ -174,3 +171,56 @@ params {
...
+            [
+                path: { "${params.outdir}/GenomeBinning/MetaBAT2/discarded_contigs" },
+                mode: 'copy',
+                pattern: '*tooShort.fa.gz'
+            ],
+            [
+                path: { "${params.outdir}/GenomeBinning/MetaBAT2/discarded_contigs" },
+                mode: 'copy',
+                pattern: '*lowDepth.fa.gz'
+            ]
```
Comment (d4straub) [inline conf/modules.config:209]: To keep it in line with the output docs and previous versions:
```suggestion
            [
                path: { "${params.outdir}/GenomeBinning/MetaBAT2/discarded" },
                mode: 'copy',
                pattern: '*tooShort.fa.gz'
            ],
            [
                path: { "${params.outdir}/GenomeBinning/MetaBAT2/discarded" },
                mode: 'copy',
                pattern: '*lowDepth.fa.gz'
            ]
```

File: conf/modules.config [hunk 2]
```diff
@@ -174,3 +171,56 @@ params {
...
+        ext.args = [
+            "-m ${params.min_contig_size}",
+            "--unbinned",
+            "--seed ${params.metabat_rng_seed}"
+        ].join(' ').trim()
+    }
+
+    withName: SPLIT_FASTA {
+        publishDir = [
+            path: { "${params.outdir}/GenomeBinning/MetaBAT2/" },
```
Comment (d4straub) [inline conf/modules.config:221]: Please add files `*.unbinned.pooled.fa.gz` & `*.unbinned.remaining.fa.gz` into folder `GenomeBinning/MetaBAT2/discarded` but `*.unbinned.[0-9]*.fa.gz` into `GenomeBinning/MetaBAT2/unbinned`

File: subworkflows/local/binning_preparation.nf [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    BOWTIE2_ASSEMBLY_ALIGN ( ch_bowtie2_input )
+    // group mappings for one assembly
+    ch_grouped_mappings = BOWTIE2_ASSEMBLY_ALIGN.out.mappings
+        .groupTuple(by: 0)
+        .map { meta, assembly, bams, bais -> [ meta, assembly[0], bams, bais ] }     // multiple symlinks to the same assembly -> use first
+
+    ch_grouped_mappings
+
```
Comment (d4straub) [inline subworkflows/local/binning_preparation.nf:48]:
```suggestion
```

**Review by d4straub — CHANGES_REQUESTED** [review]
Now I have tested with the test profile, and the output is a little unsorted yet, see previous comments.
I do think that should be sorted out first.
The general workflow seems ok to me, all tests that I made were fine (I did _not_ test group assembly, but I do not expect problems there).

File: conf/modules.config (hunk 1 above)
Comment (jfy133) [reply conf/modules.config:209]: Ok, I'll try this again for you. Like I said in the opening PR I had at some points some issues where two modules seemed to overwrite the output of each other when they saved to the same publishDir (which is why I changed the folder structure slightly), but maybe that was a conflict between syntaxes.

**Comment by jfy133** [comment]
Thanks for the review @d4straub . I'll look at the comments first half of this will.

Correct that maxbin is not included. It will come after the aDNA workflow from @maxibor 

It will be easy to insert after this restructuring. I didn't want to overload the PR as this ended up being more than I thought.

File: docs/output.md [hunk 4]
```diff
@@ -253,8 +253,8 @@ Sequencing depth per contig and sample is generated by `jgi_summarize_bam_contig
 <summary>Output files</summary>
 
 * `GenomeBinning/MetaBAT2/`
-    * `[assembler]-[sample/group].*.fa`: Genome bins retrieved from input assembly
-    * `[assembler]-[sample/group].unbinned.*.fa`: Contigs that were not binned with other contigs but considered interesting. By default, these are at least 1 Mbp (`--min_length_unbinned_contigs`) in length and at most the 100 longest contigs (`--max_unbinned_contigs`) are reported
+    * `bins/[assembler]-[sample/group].*.fa`: Genome bins retrieved from input assembly
+    * `discarded/[assembler]-[sample/group].{unbinned,lowDepth,tooShort}.*.fa`: Contigs that were not binned with other contigs but considered interesting. By default, these are at least 1 Mbp (`--min_length_unbinned_contigs`) in length and at most the 100 longest contigs (`--max_unbinned_contigs`) are reported
```
[the remark this answers is not in the document]
Comment (jfy133) [reply docs/output.md:257]: Should be fixed now

**Comment by jfy133** [comment]
@d4straub output sorted out (sorry about that, I realise looking at it I tried something else and forgot to revert back), and all other changes should be addressed :)

Like I said, MaxBin2 will come in hte next PR as I will have to change a few things more funadmentally (adding the binner to the meta tuple, for example)

File: docs/output.md [hunk 5]
```diff
@@ -253,8 +253,9 @@ Sequencing depth per contig and sample is generated by `jgi_summarize_bam_contig
 <summary>Output files</summary>
```
Comment (d4straub) [inline docs/output.md:254]: Ok, so the pipeline output is perfect now, just the output description is now redundant for  `GenomeBinning/MetaBAT2/discarded/`, in lines 258 **and** 267-269. Also, `GenomeBinning/MetaBAT2/unbinned/` is refered to in lines 257 **and** 270.

Line 262 reads `All the files and contigs in this folder will be assessed by QUAST and BUSCO.` and therefore folder `discarded` should be not included here, i.e. line 258 should be deleted.

The files `unbinned.pooled,unbinned.remaining` are now in folder discarded, so line 270 should be deleted.

I attempt to make suggestions below to solve that.
However, I could not make a suggestion for Line 262, which includes now 2 folders and therefore probably should be `All the files and contigs in those folders will be assessed by QUAST and BUSCO.`

File: docs/output.md [hunk 6]
```diff
@@ -253,8 +253,9 @@ Sequencing depth per contig and sample is generated by `jgi_summarize_bam_contig
...
 
 * `GenomeBinning/MetaBAT2/`
-    * `[assembler]-[sample/group].*.fa`: Genome bins retrieved from input assembly
-    * `[assembler]-[sample/group].unbinned.*.fa`: Contigs that were not binned with other contigs but considered interesting. By default, these are at least 1 Mbp (`--min_length_unbinned_contigs`) in length and at most the 100 longest contigs (`--max_unbinned_contigs`) are reported
+    * `bins/[assembler]-[sample/group].*.fa.gz`: Genome bins retrieved from input assembly
+    * `unbinned/[assembler]-[sample/group].unbinned.[1-9]*.fa.gz`: Contigs that were not binned with other contigs but considered interesting. By default, these are at least 1 Mbp (`--min_length_unbinned_contigs`) in length and at most the 100 longest contigs (`--max_unbinned_contigs`) are reported
+    * `discarded/[assembler]-[sample/group].{lowDepth,tooShort,unbinned,poooled,unbinned.remaining}.*.fa.gz`: Contigs that were not binned with other contigs but considered interesting. These consist of the ones that did not pass the length filtering thresholds
```
Comment (d4straub) [inline docs/output.md:258]: Those contigs are not further processed:
```suggestion
```

File: docs/output.md [hunk 7]
```diff
@@ -266,6 +267,7 @@ All the files and contigs in this folder will be assessed by QUAST and BUSCO.
 * `GenomeBinning/MetaBAT2/discarded/`
     * `*.lowDepth.fa.gz`: Low depth contigs that are filtered by MetaBat2
     * `*.tooShort.fa.gz`: Too short contigs that are filtered by MetaBat2
+* `GenomeBinning/MetaBAT2/unbinned/`
```
Comment (d4straub) [inline docs/output.md:270]: The following two files are now actually in `discarded`, so:
```suggestion
```

**Review by d4straub — COMMENTED** [review]
Ok, almost perfect, still found a confusion with the output, I think I contributed to that in my previous review! I added some suggestions to solve it. When this is done I am fine with the PR.

File: docs/output.md (hunk 5 above)
Comment (jfy133) [reply docs/output.md:254]: Updated! Hope everything shuld be clear now :)

**Review by d4straub — APPROVED** [review]
LGTM!

**Pull request merged**